_posts/2018-02-05-chaining-http-requests-in-elm.md

```diff
@@ -0,0 +1,340 @@
+- Then if we have one, we fetch the event's user full name,
+- And we map the future result of this task to the `Name` message.
+
+The cool thing here is that if anything fails along the chain, the chain stops and the error will be propagated down to the `Name` handler. No need to check errors for each operation!
+
+Now, how are we going to execute the resulting command and process the result? We need to setup the [Elm Architecture] and its good old `update` function:
+
+```haskell
+module Main exposing (main)
+
+import Html exposing (..)
+import Http
+import Json.Decode as Decode
+import Task exposing (Task)
+
+
+type alias Model =
+    { name : Maybe String
+    , error : String
+    }
+
+type Msg
+    = Name (Result String String)
+
+eventsRequest : Http.Request (List String)
+eventsRequest =
+    Http.get "https://api.github.com/events"
+        (Decode.list (Decode.at [ "actor", "login" ] Decode.string))
+
+nameRequest : String -> Http.Request String
+nameRequest login =
+    Http.get ("https://api.github.com/users/" ++ login)
+        (Decode.at [ "name" ]
+            (Decode.oneOf
+                [ Decode.string
+                , Decode.null "unspecified"
+                ]
+            )
+        )
+
+toHttpTask : Http.Request a -> Task String a
+toHttpTask request =
+    request
+        |> Http.toTask
+        |> Task.mapError toString
+
+fetchEvents : Task String (List String)
+fetchEvents =
+    toHttpTask eventsRequest
+
+fetchName : String -> Task String String
+fetchName login =
+    toHttpTask (nameRequest login)
+
+pickFirst : List String -> Task String String
+pickFirst events =
+    case List.head events of
+        Just event ->
+            Task.succeed event
+
+        Nothing ->
+            Task.fail "No events."
+
+init : ( Model, Cmd Msg )
+init =
+    { name = Nothing, error = "" }
+        ! [ fetchEvents
+                |> Task.andThen pickFirst
+                |> Task.andThen fetchName
+                |> Task.attempt Name
+          ]
+
+update : Msg -> Model -> ( Model, Cmd Msg )
+update msg model =
+    case msg of
+        Name (Ok name) ->
+            { model | name = Just name } ! []
+
+        Name (Err error) ->
+            { model | error = error } ! []
+
+view : Model -> Html Msg
+view model =
+    div []
+        [ if model.error /= "" then
+            div []
+                [ h4 [] [ text "Error encountered" ]
+                , pre [] [ text model.error ]
+                ]
+          else
+            text ""
+        , p [] [ text <| Maybe.withDefault "Fetching..." model.name ]
+        ]
+
+main =
+    Html.program
+        { init = init
+        , update = update
+        , subscriptions = always Sub.none
+        , view = view
+        }
+```
+
+That's for sure more code than with the JavaScript example, but don't forget that the Elm version renders HTML, not just logs in the console, and that the JavaScript code could be refactored to look a lot like the Elm version.
```

Comment:
Also the Elm version is fully typed and "safeguarded" against unforeseen problems (like a result returning something that's not json... not sure what would happen in the javascript example in such a case?)